File: python/massRegionHists.py
from hists import *
from eventViewHists import *

class massRegionHists:
    def __init__(self, outFile, directory):
        #once diJets are constructed, one can ask what regions of the 2D plane defined by (dijet1 mass, dijet2 mass) a given pair of dijets lands in.
        #example: make plots of dijet masses for events where both dijets are consistent with the Z boson mass.

        #plot all dijet pairings
        outFile.mkdir(directory+"/inclusive")
        self.inclusive = eventViewHists(outFile, directory+"/inclusive")

        #Signal Region (SR)
        #only plot dijet pairings that are consistent with BB->(jj)(jj) where B is Z or H
        outFile.mkdir(directory+"/ZZSR")
        self.ZZSR        = eventViewHists(outFile, directory+"/ZZSR", True)
        outFile.mkdir(directory+"/ZHSR")
        self.ZHSR        = eventViewHists(outFile, directory+"/ZHSR", True)
        outFile.mkdir(directory+"/HHSR")
        self.HHSR        = eventViewHists(outFile, directory+"/HHSR", True)

        #Control Region (CR)
        outFile.mkdir(directory+"/ZZCR")
        self.ZZCR        = eventViewHists(outFile, directory+"/ZZCR", True)
        outFile.mkdir(directory+"/ZHCR")
        self.ZHCR        = eventViewHists(outFile, directory+"/ZHCR", True)
        outFile.mkdir(directory+"/HHCR")
        self.HHCR        = eventViewHists(outFile, directory+"/HHCR", True)

        #Sideband (SB)
        outFile.mkdir(directory+"/ZZSB")
        self.ZZSB        = eventViewHists(outFile, directory+"/ZZSB", True)
        outFile.mkdir(directory+"/ZHSB")
        self.ZHSB        = eventViewHists(outFile, directory+"/ZHSB", True)
        outFile.mkdir(directory+"/HHSB")
        self.HHSB        = eventViewHists(outFile, directory+"/HHSB", True)
        
    def Fill(self, view, weight=1, event=None):
        self.inclusive.Fill(view, weight)
        if view.ZZSR: self.ZZSR.Fill(view, weight, event)
        if view.ZHSR: self.ZHSR.Fill(view, weight, event)
        if view.HHSR: self.HHSR.Fill(view, weight, event)

        if view.ZZCR: self.ZZCR.Fill(view, weight, event)
        if view.ZHCR: self.ZHCR.Fill(view, weight, event)
        if view.HHCR: self.HHCR.Fill(view, weight, event)

        if view.ZZSB: self.ZZSB.Fill(view, weight, event)
        if view.ZHSB: self.ZHSB.Fill(view, weight, event)
        if view.HHSB: self.HHSB.Fill(view, weight, event)

    def Write(self, outFile=None):
        #self.thisDir.cd()
        self.inclusive.Write()
        self.ZZSR.Write()
        self.ZHSR.Write()
        self.HHSR.Write()

        self.ZZCR.Write()
        self.ZHCR.Write()
        self.HHCR.Write()

        self.ZZSB.Write()
        self.ZHSB.Write()
        self.HHSB.Write()

File: NtupleAna/src/massRegionHists.cc
#include "ZZ4b/NtupleAna/interface/massRegionHists.h"

using namespace NtupleAna;

massRegionHists::massRegionHists(std::string name, fwlite::TFileService& fs, bool isMC, bool _blind) {
  dir = fs.mkdir(name);
  blind = _blind;

  inclusive = new viewHists(name+"/inclusive", fs, isMC);

  ZZSR      = new viewHists(name+"/ZZSR",      fs, isMC);
  ZHSR      = new viewHists(name+"/ZHSR",      fs, isMC);

  ZZCR      = new viewHists(name+"/ZZCR",      fs, isMC);
  ZHCR      = new viewHists(name+"/ZHCR",      fs, isMC);

  ZZSB      = new viewHists(name+"/ZZSB",      fs, isMC);
  ZHSB      = new viewHists(name+"/ZHSB",      fs, isMC);

  ZH        = new viewHists(name+"/ZH",        fs, isMC);

  ZH_SvB_high = new viewHists(name+"/ZH_SvB_high", fs, isMC);
  ZH_SvB_low  = new viewHists(name+"/ZH_SvB_low",  fs, isMC);

} 

void massRegionHists::Fill(eventData* event, std::unique_ptr<eventView> &view){
  if(blind && (view->ZZSR || view->ZHSR || view->HHSR)) return;
  
  inclusive->Fill(event, view);
  if(view->ZZSR) ZZSR->Fill(event, view);
  if(view->ZHSR) ZHSR->Fill(event, view);

  if(view->ZZCR) ZZCR->Fill(event, view);
  if(view->ZHCR) ZHCR->Fill(event, view);

  if(view->ZZSB) ZZSB->Fill(event, view);
  if(view->ZHSB) ZHSB->Fill(event, view);

  if(view->ZHSB || view->ZHCR || view->ZHSR){
    ZH->Fill(event, view);
    if(event->ZHvsBackgroundClassifier > 0.5)
      ZH_SvB_high->Fill(event, view);
    else
      ZH_SvB_low ->Fill(event, view);
  }

  return;
}

massRegionHists::~massRegionHists(){} 


